packages/controls/src/controls/shape-v2/shape-v2.ts

```diff
@@ -0,0 +1,322 @@
+      data[ControlDataTypeKey] === Definition.v1DataType
+    ) {
+      return mapValues(this.keyDefs, (def, key) =>
+        def.copyData((data as ShapeV2DataTypeV1<C>).value[key], context),
+      )
+    }
+
+    return mapValues(this.keyDefs, (def, key) =>
+      def.copyData((data as ShapeV2DataTypeV0<C>)[key], context),
+    )
+  }
+
+  getTranslatableData(data: DataType<C>): Data {
+    if (data == null) return null
+
+    if (
+      ControlDataTypeKey in data &&
+      data[ControlDataTypeKey] === Definition.v1DataType
+    ) {
+      return mapValues(this.keyDefs, (def, key) =>
+        def.getTranslatableData((data as ShapeV2DataTypeV1<C>).value[key]),
+      )
+    }
+
+    return mapValues(this.keyDefs, (def, key) =>
+      def.getTranslatableData((data as ShapeV2DataTypeV0<C>)[key]),
+    )
+  }
+
+  mergeTranslatedData(
+    data: DataType<C>,
+    translatedData: Record<string, DataType<C>>,
+    context: MergeTranslatableDataContext,
+  ): Data {
+    if (translatedData == null) return data
+
+    if (
+      ControlDataTypeKey in data &&
+      data[ControlDataTypeKey] === Definition.v1DataType
+    ) {
+      return mapValues(this.keyDefs, (def, key) =>
+        def.mergeTranslatedData(
+          (data as ShapeV2DataTypeV1<C>).value[key],
+          translatedData[key],
+          context,
+        ),
+      )
+    }
+
+    return mapValues(this.keyDefs, (def, key) =>
+      def.mergeTranslatedData(
+        (data as ShapeV2DataTypeV0<C>)[key],
+        translatedData[key],
+        context,
+      ),
+    )
+  }
+
+  createInstance(sendMessage: SendMessage): InstanceType<C> {
+    return new ShapeV2Control(this, sendMessage)
+  }
+
+  serialize(): [SerializedRecord, Transferable[]] {
+    return serialize(this.config, {
+      type: Definition.type,
+    })
+  }
+
+  introspect<R>(data: DataType<C> | undefined, target: IntrospectionTarget<R>) {
+    if (data == null) return []
+
+    if (
+      ControlDataTypeKey in data &&
+      data[ControlDataTypeKey] === Definition.v1DataType
+    ) {
+      return Object.entries(this.keyDefs).flatMap(
+        ([key, def]) =>
+          def.introspect((data as ShapeV2DataTypeV1<C>).value[key], target) ??
+          [],
+      )
+    }
+
+    return Object.entries(this.keyDefs).flatMap(
+      ([key, def]) =>
+        def.introspect((data as ShapeV2DataTypeV0<C>)[key], target) ?? [],
+    )
+  }
+}
+
+type UserConfig<
+  D extends KeyDefinitions,
+  L extends Config['layout'],
+> = Config<D> & {
+  layout?: L
+}
+
+type NormedConfig<
+  D extends KeyDefinitions,
+  L extends Config['layout'] = Config['layout'],
+> = undefined extends L
+  ? Config<D>
+  : Pick<Config<D>, 'label' | 'type'> & {
+      layout: L
+    }
+
+export class ShapeV2Definition<
+  C extends Config = Config,
+> extends Definition<C> {}
+
+export function ShapeV2<D extends KeyDefinitions, L extends Config['layout']>(
+  config: UserConfig<D, L>,
+): ShapeV2Definition<NormedConfig<D, L>> {
```

Comment:
Since our data type doesn't really depend on the `layout`, I wonder if we should bother with the complexity of trying to infer the specific type of the layout passed here.

Reply:
Removed